apps/dashboard/src/hooks/use-form-autosave.ts

```diff
@@ -0,0 +1,33 @@
+import { useCallback, useRef } from 'react';
+import { FieldValues, SubmitHandler, UseFormReturn, useWatch } from 'react-hook-form';
+import debounce from 'debounce';
+import useDeepCompareEffect from 'use-deep-compare-effect';
+
+export const useFormAutoSave = <T extends FieldValues>({
+  onSubmit,
+  form,
+}: {
+  onSubmit: SubmitHandler<T>;
+  form: UseFormReturn<T>;
+}) => {
+  const onSubmitRef = useRef<SubmitHandler<T>>(onSubmit);
+  onSubmitRef.current = onSubmit;
+  const { formState, control, handleSubmit } = form;
+
+  const watchedData = useWatch<T>({
+    control,
+  });
+
+  const debouncedSave = useCallback<() => void>(
+    debounce(() => {
+      handleSubmit(onSubmitRef.current)();
+    }, 1000),
+    [handleSubmit]
+  );
+
+  useDeepCompareEffect(() => {
+    if (formState.isDirty) {
+      debouncedSave();
+    }
+  }, [watchedData]);
```

Comment:
Basically, we watch all the form data changes and compare it deeply with the previous form-watched data between re-renders. If the form data was changed and the form is dirty, we call debounced save, which calls the react hook form `handleSubmit` that triggers validation. If the data is valid, then the `onSubmit` will be called.

Reply:
Maybe additionally we can show the sonner when there are validation errors?